### ISSUE TITLE
A11Y Android: using a screen reader in an edit box says all the time to substituted

### ISSUE THREAD

**Issue by ericdq** [laurent22/joplin#13191]
On Android using TalkBack, when typing in any edit box, be it the search box, the note title box, the body box... In all of them, when typing any letter it says substituted.

If for example I write hello, it says it in every letter I press adding the next one I write. This is annoying in short sentences but in long sentences it is unbearable.

There are times when this behaviour stops happening.

I don't know what kind of input is being used or if you have any options added underneath for screen readers. It would be good to check this point, as in other edit boxes of other applications this behaviour does not occur.

**Comment by personalizedrefrigerator** [comment]
I can reproduce this with an Android 16 emulator with the GBoard keyboard.

Notes:

- This seems similar to [an upstream React Native issue](https://github.com/facebook/react-native/issues/48233) that was closed as "unsupported version".
- I can't reproduce this with the note body.
   - For me, the plain text editor is disabled (which does mean that I'm [affected by this bug](https://issues.chromium.org/u/1/issues/433829629), since the editor uses a WebView).

**Comment by ericdq** [comment]
thanks for attaching the link. It explains much better what is happening to me. I have tried other applications that use webview, like google search. Or opening an email with a link and there activating the search box and opening the google page to do a search on it but with the component loaded and there it reads it fine.

I don't know if it could be some communication or some library that is failing in the communication between both services.

When you say that the plain text editor is disabled do you mean that you use the rich text editor @personalizedrefrigerator ?

**Comment by personalizedrefrigerator** [comment]
> I have tried other applications that use webview, like google search. Or opening an email with a link and there activating the search box and opening the google page to do a search on it but with the component loaded and there it reads it fine.

For me, it was the WebView-based Markdown editor that read correctly ☺ . The other inputs (e.g. the title input) weren't working correctly. These other inputs use React Native [TextInput](https://reactnative.dev/docs/0.79/textinput)s, which I think is related to the issue.

> When you say that the plain text editor is disabled do you mean that you use the rich text editor @personalizedrefrigerator ?

For me, the issue did not occur with the Markdown editor.

The plain text editor refers to a compatibility option that enables a Markdown editor with fewer features. In particular, in General > Note, there's a "Use the plain text editor" setting. If enabled, Joplin's note editor uses a React Native `TextInput`, rather than the WebView-based CodeMirror input. Since WebView-based editors usually [have trouble with the navigate-by-lines and navigate-by-paragraphs](https://github.com/codemirror/dev/issues/1556) on Android, enabling the "Use the plain text editor" setting was a potential workaround that would allow navigate-by-line and navigate-by-paragraph to work within Joplin.

**Comment by ericdq** [comment]
sorry for the confusion. I was indeed using the markdown editor as well.

In the body of the note it happens with fewer letters but it can also be reproduced. I would say when typing the first three characters.

Maybe if you type very fast you don't notice it. But if you type, let some time pass until you press the next letter, that's when it happens.

I don't know if it's because I tested this in the previous update, but I would say that it used to happen at least until you got to write about 7 words.

I have continued writing and although it doesn't say substituted, every time I write a letter or press space it reads everything I have written until a moment comes, as I said above, it seems that when I reach a certain number of characters it stops doing it. I have even tried to add a new line and it keeps happening, so I think it may be in some code, or when sizing the writing box.

**Comment by personalizedrefrigerator** [comment]
So far, I'm still unable to reproduce the issue in the Markdown editor. This might be because I'm using an Android emulator. However, I should be able to test on a physical device in several hours.

**Additional information**: I'm testing on an Android 16 emulator with TalkBack (version 16.0.0.738667889 of the Android Accessibility Suite) and the GBoard keyboard. My system WebView version is 139.0.7258.158 (com.google.android.webview). 

**Follow-up question:** There's an online version of the mobile app at [app.joplincloud.com](https://app.joplincloud.com/) (which uses `react-native-web`). Are you able to reproduce the Markdown editor issue with the app.joplincloud.com version of the app? Alternatively, the Markdown editor uses the [CodeMirror editor](https://codemirror.net/try/), which has a [few online demos](https://codemirror.net/try/). Can you reproduce the issue in one of these editors?
- As commented above, I should be able to test this myself on a physical Android device (Android 13) in a few hours.

**Comment by ericdq** [comment]
I have the same version of TalkBack, although I'm using Android 13 on a Pixel 4A.

I've tested it on app.goplincloud.com and here we have a different behaviour. The note title reads correctly, as in other edit boxes I've mentioned before. However, in the body, with the Markdown editor, the same thing still happens, it doesn't say it is subtracted, but with every character typed it reads everything in the edit box.

I don't know how I could find out the webview version on my system. I also use Gboard.

In the CodeMirror box, this behaviour doesn't seem to happen either.

**Comment by personalizedrefrigerator** [comment]
Thank you for testing!

I plan to:
- Focus on the title/search/settings input issue for now, since I can reproduce that one locally. 
- Try to reproduce the Markdown editor issue with an Android 13 device.

**Comment by personalizedrefrigerator** [comment]
### Update: The title/search/settings input bug seems to be an upstream issue

I'm able to reproduce this with controlled `TextInput`s using the React Native `main` branch. I've created an [issue reproducer](https://github.com/facebook/react-native/pull/53764) for the [existing closed/stale upstream issue](https://github.com/facebook/react-native/issues/48233).

For now, possible workarounds in the Joplin codebase include:
- Switching all `TextInput`s to uncontrolled inputs if possible.
   - This would involve using React Native's [`defaultValue`](https://reactnative.dev/docs/textinput#defaultvalue) prop instead of [`value`](https://reactnative.dev/docs/textinput#value).
   - This might not work in cases where the `TextInput`s need to be updated by the app after being created (e.g. with the note title input). 
- Try to fix the issue by patching React Native.

Since the Markdown editor doesn't use a React Native `TextInput`, its issue probably has a different source.

**Comment by personalizedrefrigerator** [comment]
I've tested this on Android 13 with TalkBack 13.5.02.3. With this setup, I can't reproduce either issue (neither the TextInput issue nor the Markdown editor issue). I suspect that this issue is related to a recent change in TalkBack, conflicting with how React Native handles [controlled inputs](https://reactnative.dev/docs/0.79/textinput#value).

**Comment by ericdq** [comment]
Thank you very much for all your research. I read this week that they are going to release TalkBack version 16.1. As soon as I receive it, I will test to see if the problem still occurs.

**Comment by personalizedrefrigerator** [comment]
I've opened an [issue report for this on the Google issue tracker](https://issuetracker.google.com/issues/445102895).

**Comment by personalizedrefrigerator** [comment]
According to the upstream issue report, this has been resolved in TalkBack 17.0.1.

**Comment by ericdq** [comment]
Thank you for your attention and for following up on the issue. I hadn’t commented here because version 17.1 of TalkBack hadn’t arrived on my phone yet, but I think we could close this thread based on the response given in the other thread.
